experimental/lib/Analysis/GsaAnalysis.cpp

```diff
@@ -0,0 +1,485 @@
+    // value and return it
+    auto getRestrictedExpression =
+        [&](bool value) -> boolean::BoolExpression * {
+      auto *expr = expression.first->deepCopy();
+      boolean::restrict(expr, conditionToUse, value);
+      return expr->boolMinimize();
+    };
+
+    // Substitute a condition in the expression both with value true and false
+    auto *exprTrue = getRestrictedExpression(true);
+    auto *exprFalse = getRestrictedExpression(false);
+
+    // One of the two expressions above will be zero: add the input to the
+    // corresponding list
+    if (exprTrue->toString() != "0")
+      conditionsTrueExpressions.emplace_back(exprTrue, expression.second);
+    if (exprFalse->toString() != "0")
+      conditionsFalseExpressions.emplace_back(exprFalse, expression.second);
+  }
+
+  SmallVector<GateInput *> operandsGamma(2);
+
+  // If the number of non-null expressions obtained with cofactor = X is greater
+  // than 1, then the expansions must be done again over those inputs with a new
+  // cofactor. The resulting gamma is going to be used as X input of this
+  // gamma.
+  //
+  // If only one element is present, that same element is used as input of the
+  // gamma.
+  //
+  // If no elements are present, then that input will never be used, and it is
+  // considered as empty.
+  auto setGammaOperand = [&](int input, ListExpressionsPerGate &list) -> void {
+    if (list.size() > 1) {
+      auto *gamma = expandGammaTree(list, conditions, originalPhi);
+      operandsGamma[input] = new GateInput(gamma);
+    } else if (list.size() == 1) {
+      operandsGamma[input] = list[0].second;
+    } else {
+      operandsGamma[input] = new GateInput();
+    }
+  };
+
+  setGammaOperand(1, conditionsTrueExpressions);
+  setGammaOperand(0, conditionsFalseExpressions);
+
+  // Create a new gamma and add it to the list of phis for the original basic
+  // block.
+  auto *newGate =
+      new Gate(originalPhi->result, operandsGamma, GateType::GammaGate,
+               ++uniqueGateIndex, conditionToUse);
+  gateList[originalPhi->getBlock()].push_back(newGate);
+
+  return newGate;
+}
+
+template <typename FunctionType>
+void GsaAnalysis<FunctionType>::identifyAllGates(FunctionType &funcOp) {
+
+  // This function works in two steps. First, all the block arguments in the IR
+  // are converted into PHIs, taking care or properly extracting the information
+  // about the producers of the operands. Then, the phis ar converted either to
+  // GAMMAs or MUs.
+
+  // The input of a phi might be another . This is the case when the
+  // input of a phi is a block argument from a block with index different
+  // from 0. In this situation, we mark the phi input as `missing`, and we
+  // store the necessary information (block argument) to later reconstruct the
+  // relationship.
+  struct MissingPhi {
+
+    // Which input is missing
+    GateInput *pi;
+
+    // Related block argument
+    BlockArgument ba;
```

Comment:
`ba` is not an immediately understandable or commonly used name for a `BlockArgument`, use `arg` or `blockArg` instead.